core/src/main/kotlin/org/opensearch/alerting/core/model/ClusterMetricsInput.kt

```diff
@@ -0,0 +1,334 @@
+            throw IllegalArgumentException("The API requires path parameters.")
+        if (!apiType.supportsPathParams && pathParams.isNotEmpty())
+            throw IllegalArgumentException("The API does not use path parameters.")
+
+        return pathParams
+    }
+
+    /**
+     * Examines the path of a [ClusterMetricsInput] to determine which API is being called.
+     * @param uriPath The path to examine.
+     * @return The [ClusterMetricType] associated with the [ClusterMetricsInput] monitor.
+     * @throws IllegalArgumentException when the API to call cannot be determined from the URI.
+     */
+    private fun findApiType(uriPath: String): ClusterMetricType {
+        var apiType = ClusterMetricType.BLANK
+        ClusterMetricType.values()
+            .filter { option -> option != ClusterMetricType.BLANK }
+            .forEach { option ->
+                if (uriPath.startsWith(option.prependPath) || uriPath.startsWith(option.defaultPath))
+                    apiType = option
+            }
+        if (apiType.isBlank())
+            throw IllegalArgumentException("The API could not be determined from the provided URI.")
+        return apiType
+    }
+
+    /**
+     * Constructs a [URI] from the supported scheme, host, and port, and the provided [path], and [pathParams].
+     * @return The constructed [URI].
+     */
+    private fun constructUrlFromInputs(): URI {
+        val uriBuilder = URIBuilder()
+            .setScheme(SUPPORTED_SCHEME)
+            .setHost(SUPPORTED_HOST)
+            .setPort(SUPPORTED_PORT)
+            .setPath(path + pathParams)
+        return uriBuilder.build()
+    }
+
+    /**
+     * If [url] field is empty, populates it with [constructedUri].
+     * If [path] and [pathParams] are empty, populates them with values from [url].
+     */
+    private fun parseEmptyFields() {
+        if (pathParams.isEmpty())
+            pathParams = this.parsePathParams()
+        if (path.isEmpty())
+            path = if (pathParams.isEmpty()) clusterMetricType.defaultPath else clusterMetricType.prependPath
+        if (url.isEmpty())
+            url = constructedUri.toString()
+    }
+
+    /**
+     * Helper function to confirm at least [url], or required URI component fields are defined.
+     * @return TRUE if at least either [url] or the other components are provided; otherwise FALSE.
+     */
+    private fun validateFields(): Boolean {
+        return url.isNotEmpty() || validateFieldsNotEmpty()
+    }
+
+    /**
+     * Confirms that required URI component fields are defined.
+     * Only validating path for now, as that's the only required field.
+     * @return TRUE if all those fields are defined; otherwise FALSE.
+     */
+    private fun validateFieldsNotEmpty(): Boolean {
+        return path.isNotEmpty()
+    }
+
+    /**
+     * An enum class to quickly reference various supported API.
+     */
+    enum class ClusterMetricType(
```

Comment:
This logic can be replaced with regex validations and in future releases we may want to keep it more generic so that if path requires some placeholders, users can configure them without changing code. For e.g. node stats api supports adding node id or cluster metric name as part of url path.

Reply:
Created this issue to track this feedback https://github.com/opensearch-project/alerting/issues/341

Reply:
Manual testing involved creating a user with all permissions except `cluster:monitor` permissions. This prevents that user from running API like `GET /_cluster/heath`. The `Preview query` button for a `ClusterMetrics` monitor defined by that user fails to execute. In addition, execution of the monitor results in alerts with the following error
```
Error: Failed fetching inputs: OpenSearchSecurityException[no permissions for [cluster:monitor/health] and associated roles [noClusterApi, own_index]]
```

Also, I created a user without access to the `opensearch_dashboards_sample_data_ecommerce` index. Attempting to preview the response for a `/_cluster/heath/opensearch_dashboards_sample_data_ecommerce` monitor returns the same information as if that user ran `GET /_cluster/heath/opensearch_dashboards_sample_data_ecommerce` through DevTools. Attempts for that user to preview the response for `_cat/recovery/opensearch_dashboards_sample_data_ecommerce` fail to execute, and attempts to create the monitor display a popup error that says 
```
[alerting_exception] User doesn't have read permissions for one or more configured index []
```